### PR TITLE
Add Py_mod_gil slot to C extension modules

### PR DESCRIPTION
**Pull request by lysnikolaou**
x-ref #59057.

The `Py_mod_gil` slot is needed so that the GIL is not automatically enabled when these C extension modules get imported.

- I checked `pandas_datetime` for thread-safety and it appears to be okay.
- `pandas_parser` is not thread-safe. A discussion is needed on whether we should care about it though, since the `c` engine to `read_csv` and friends is already documented as being thread-unsafe and the thread-safe `pyarrow` alternative is already there.

### CODE REVIEW

**Comment by WillAyd** [comment]
Is this slot only available in 3.13?

> * `pandas_parser` is not thread-safe. A discussion is needed on whether we should care about it though,

Do you know what makes it not thread unsafe? I'm surprised given how small of a module it is.

Definitely more of a nice-to-have on the parser though. Understood its probably a very large undertaking to make the CSV reader overall multi-threaded, and most likely not even worth it with pyarrow

File: pandas/_libs/src/datetime/pd_datetime.c [hunk 1]
```diff
@@ -245,7 +245,10 @@ static int pandas_datetime_exec(PyObject *Py_UNUSED(module)) {
 }
 
 static PyModuleDef_Slot pandas_datetime_slots[] = {
-    {Py_mod_exec, pandas_datetime_exec}, {0, NULL}};
+    {Py_mod_exec, pandas_datetime_exec},
+    {Py_mod_gil, Py_MOD_GIL_NOT_USED},
```
Comment (colesbury) [inline pandas/_libs/src/datetime/pd_datetime.c:249]: As @WillAyd mentioned, this is new in 3.13:

```
#if PY_VERSION_HEX >= 0x030D0000
    {Py_mod_gil, Py_MOD_GIL_NOT_USED},
#endif
```

https://docs.python.org/3.14/howto/free-threading-extensions.html#multi-phase-initialization

Comment (lysnikolaou) [reply pandas/_libs/src/datetime/pd_datetime.c:250]: Thanks for this. Friday was not a good day, apparently. 🤕 

One question: In the [free-threaded repo's README](https://github.com/Quansight-Labs/free-threaded-compatibility?tab=readme-ov-file#running-python-with-the-gil-disabled) we recomment guarding with `#ifdef Py_GIL_DISABLED`. Should we change that?

**Comment by lysnikolaou** [comment]
> Do you know what makes it not thread unsafe? I'm surprised given how small of a module it is.

I mean that it's thread-unsafe in the sense that there's global state in the `parser_t` struct and there's unsafe reads/writes to it. As far as I understand, `read_csv` only ever creates one instance of it though, which means that running `read_csv` in multiple threads is safe.

**Comment by WillAyd** [comment]
Sorry for the typo - I mean "not thread safe" but I think you got what I was saying. And cool thanks for the response - agree with you on that

File: pandas/_libs/src/datetime/pd_datetime.c (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/src/datetime/pd_datetime.c:250]: I haven't kept tabs on the CPython implementation but seems like what the free-threaded repo suggests might be better than checking the version hex - there will be both GIL and GIL-less Python compilations for a few versions right?

Comment (lysnikolaou) [reply pandas/_libs/src/datetime/pd_datetime.c:250]: @WillAyd That's correct, however `Py_mod_gil` is just ignored in 3.13+ non-free-threaded builds. Effectively, both guard lead to correct behavior.

Comment (WillAyd) [reply pandas/_libs/src/datetime/pd_datetime.c:250]: As a matter of style I think checking for `Py_GIL_DISABLED` would make more sense if you were evaluating whether that was true/false, not just checking if the symbol was defined. 

So happy with what we have here, but probably worth asking for input from other projects for the free-threaded repo standard.

We can merge this as is for now and follow up with a different approach if a more canonical one becomes clearer

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @lysnikolaou

**Pull request merged**